File: apps/server/scripts/import-resume-chunks.mjs
#!/usr/bin/env node

/**
 * 从 Milvus 导出的 JSONL 导入到 SQLite rag_documents 和 rag_chunks 表。
 *
 * 使用方式:
 *   pnpm --filter @my-resume/server rag:import:resume
 *   pnpm --filter @my-resume/server rag:import:resume --file .data/resume_profile_chunks.jsonl
 *
 * 每行 JSON 结构:
 *   { id, section, subsection_key, subsection_title, entity_type,
 *     content, vector, tags, source_id, locale, chunk_index, chunk_count }
 */

import { execFileSync } from 'node:child_process'
import { createHash } from 'node:crypto'
import { existsSync } from 'node:fs'
import { createReadStream } from 'node:fs'
import path from 'node:path'
import readline from 'node:readline'
import { fileURLToPath } from 'node:url'

const __filename = fileURLToPath(import.meta.url)
const SCRIPT_DIR = path.dirname(__filename)
const REPO_ROOT = path.resolve(SCRIPT_DIR, '../../..')
const DB_PATH = path.join(REPO_ROOT, '.data', 'my-resume.db')
const DEFAULT_INPUT = path.join(REPO_ROOT, '.data', 'resume_profile_chunks.jsonl')

// ── helpers ──────────────────────────────────────────────────────────

function parseArgs(argv) {
  const args = { file: '' }
  for (let i = 2; i < argv.length; i++) {
    if (argv[i] === '--file') {
      args.file = argv[++i]
    }
  }
  return args
}

function computeHash(content) {
  return createHash('sha256').update(content).digest('hex')
}

function escapeSql(value) {
  return String(value).replace(/'/g, "''")
}

/** 执行一条 SQL，不回滚失败 */
function runSql(sql) {
  try {
    execFileSync('sqlite3', [DB_PATH], {
      input: sql + ';\n',
      stdio: ['pipe', 'ignore', 'pipe'],
      timeout: 5000,
    })
    return true
  } catch (error) {
    const stderr = error.stderr ? error.stderr.toString() : ''
    if (stderr && !stderr.includes('UNIQUE constraint')) {
      console.error(`  sqlite3 error: ${stderr.trim()}`)
    }
    return false
  }
}

async function* readJsonlLines(filePath) {
  const stream = createReadStream(filePath, 'utf8')
  const rl = readline.createInterface({ input: stream, crlfDelay: Infinity })

  for await (const line of rl) {
    const trimmed = line.trim()
    if (!trimmed) continue
    try {
      yield JSON.parse(trimmed)
    } catch (error) {
      console.error(`WARNING: skip invalid JSON: ${error.message}`)
    }
  }
}

// ── main ─────────────────────────────────────────────────────────────

async function main() {
  const args = parseArgs(process.argv)
  const inputPath = args.file || DEFAULT_INPUT

  if (!existsSync(inputPath)) {
    console.error(`Input file not found: ${inputPath}`)
    process.exit(1)
  }

  console.log(`Reading: ${inputPath}`)

  // 1. 解析所有 chunk
  /** @type {Array<Record<string, unknown>>} */
  const chunks = []

  for await (const record of readJsonlLines(inputPath)) {
    chunks.push(record)
  }

  console.log(`Parsed ${chunks.length} chunks`)

  if (chunks.length === 0) {
    console.error('No chunks to import.')
    process.exit(1)
  }

  // 2. 构建文档记录
  const sourceId = String(chunks[0].source_id || 'fuyinsheng-resume-zh')
  const locale = String(chunks[0].locale || 'zh')
  const docId = computeHash(`resume_core:published:${sourceId}:${locale}`)
  const now = Date.now()
  const contentHash = computeHash(
    chunks.map((c) => String(c.content || '')).join('\n'),
  )

  console.log(`Target DB: ${DB_PATH}`)

  if (!existsSync(DB_PATH)) {
    console.error(`DB file does not exist: ${DB_PATH}`)
    process.exit(1)
  }

  // 3. Upsert 文档 & 清除旧 chunk
  const metadataJson = escapeSql(JSON.stringify({
    importedFrom: 'milvus',
    totalChunks: chunks.length,
    importedAt: new Date().toISOString(),
  }))

  runSql(`INSERT OR REPLACE INTO rag_documents
    (id, source_type, source_scope, source_id, source_version,
     locale, title, content_hash, metadata_json, created_at, updated_at)
    VALUES ('${docId}', 'resume_core', 'published', '${escapeSql(sourceId)}',
     'milvus-import:${new Date().toISOString().slice(0, 10)}', '${locale}',
     '${locale === "zh" ? "简历核心数据" : "Resume core data"} - ${escapeSql(sourceId)}', '${contentHash}', '${metadataJson}', ${now}, ${now})`)

  runSql(`DELETE FROM rag_chunks WHERE document_id = '${docId}'`)

  // 4. 逐条插入
  let inserted = 0

  for (let i = 0; i < chunks.length; i++) {
    const chunk = chunks[i]
    const embedding = Array.isArray(chunk.vector) ? chunk.vector : []
    const metaJson = escapeSql(JSON.stringify({
      source_id: String(chunk.source_id || ''),
      locale: String(chunk.locale || 'zh'),
      subsection_key: String(chunk.subsection_key || ''),
      subsection_title: String(chunk.subsection_title || ''),
      entity_type: String(chunk.entity_type || ''),
      tags: Array.isArray(chunk.tags) ? chunk.tags : [],
    }))

    const sql = `INSERT OR REPLACE INTO rag_chunks
      (id, document_id, chunk_index, section, content,
       content_hash, embedding_json, metadata_json,
       created_at, updated_at)
      VALUES ('${escapeSql(String(chunk.id || ''))}', '${docId}', ${i},
       '${escapeSql(String(chunk.section || chunk.subsection_key || ''))}',
       '${escapeSql(String(chunk.content || ''))}',
       '${computeHash(String(chunk.content || ''))}',
       '${escapeSql(JSON.stringify(embedding))}',
       '${metaJson}',
       ${now}, ${now})`

    if (runSql(sql)) {
      inserted += 1
    } else {
      console.error(`  skip chunk ${i} (${String(chunk.id || '').slice(0, 40)})`)
    }
  }

  console.log(`Inserted ${inserted}/${chunks.length} chunks`)
  console.log('Done.')
}

main().catch((error) => {
  console.error(error)
  process.exit(1)
})

File: apps/server/scripts/import-resume-chunks.py
#!/usr/bin/env python3
"""
从 Milvus 导出的 JSONL 导入到 SQLite rag_documents 和 rag_chunks 表。

使用方式:
  pnpm --filter @my-resume/server rag:import:resume
  pnpm --filter @my-resume/server rag:import:resume --file .data/resume_profile_chunks.jsonl

每行 JSON 结构:
  { id, section, subsection_key, subsection_title, entity_type,
    content, vector, tags, source_id, locale, chunk_index, chunk_count }
"""
import hashlib
import json
import os
import sys
import time
from pathlib import Path

REPO_ROOT = Path(__file__).resolve().parents[3]
DB_PATH = REPO_ROOT / ".data" / "my-resume.db"
DEFAULT_INPUT = REPO_ROOT / ".data" / "resume_profile_chunks.jsonl"


def resolve_input_path() -> Path:
    for i, arg in enumerate(sys.argv):
        if arg == "--file" and i + 1 < len(sys.argv):
            return Path(sys.argv[i + 1])
    return DEFAULT_INPUT


def compute_hash(content: str) -> str:
    return hashlib.sha256(content.encode()).hexdigest()


def main():
    input_path = resolve_input_path()

    if not input_path.exists():
        print(f"Input file not found: {input_path}", file=sys.stderr)
        sys.exit(1)

    print(f"Reading: {input_path}")
    chunks = []

    with open(input_path, encoding="utf-8") as f:
        for line_num, line in enumerate(f, 1):
            line = line.strip()
            if not line:
                continue
            try:
                record = json.loads(line)
                chunks.append(record)
            except json.JSONDecodeError as e:
                print(f"WARNING: skip invalid JSON at line {line_num}: {e}")

    print(f"Parsed {len(chunks)} chunks")

    if not chunks:
        print("No chunks to import.", file=sys.stderr)
        sys.exit(1)

    # 构建文档记录
    source_id = str(chunks[0].get("source_id", "fuyinsheng-resume-zh"))
    locale = str(chunks[0].get("locale", "zh"))
    doc_id = compute_hash(f"resume_core:published:{source_id}:{locale}")
    now = int(time.time() * 1000)
    content_hash = compute_hash(
        "".join(str(c.get("content", "")) for c in chunks)
    )

    db_path = str(DB_PATH)
    print(f"Target DB: {db_path}")

    if not DB_PATH.exists():
        print(f"DB file does not exist yet: {db_path}")
        sys.exit(1)

    import sqlite3
    db = sqlite3.connect(db_path)

    # 插入或替换文档记录
    db.execute(
        """
        INSERT OR REPLACE INTO rag_documents
          (id, source_type, source_scope, source_id, source_version,
           locale, title, content_hash, metadata_json, created_at, updated_at)
        VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?)
        """,
        (
            doc_id,
            "resume_core",
            "published",
            source_id,
            f"milvus-import:{time.strftime('%Y-%m-%d')}",
            locale,
            "付寅生简历核心数据",
            content_hash,
            json.dumps({
                "importedFrom": "milvus",
                "totalChunks": len(chunks),
                "importedAt": time.strftime("%Y-%m-%dT%H:%M:%SZ", time.gmtime()),
            }, ensure_ascii=False),
            now,
            now,
        ),
    )

    # 清除旧 chunk
    db.execute("DELETE FROM rag_chunks WHERE document_id = ?", (doc_id,))

    # 按顺序插入新 chunk
    inserted = 0
    for i, chunk in enumerate(chunks):
        try:
            db.execute(
                """
                INSERT INTO rag_chunks
                  (id, document_id, chunk_index, section, content,
                   content_hash, embedding_json, metadata_json,
                   created_at, updated_at)
                VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?, ?)
                """,
                (
                    str(chunk.get("id", "")),
                    doc_id,
                    i,
                    str(chunk.get("section", chunk.get("subsection_key", ""))),
                    str(chunk.get("content", "")),
                    compute_hash(str(chunk.get("content", ""))),
                    json.dumps(chunk.get("vector", [])),
                    json.dumps({
                        "source_id": str(chunk.get("source_id", "")),
                        "locale": str(chunk.get("locale", "zh")),
                        "subsection_key": str(chunk.get("subsection_key", "")),
                        "subsection_title": str(chunk.get("subsection_title", "")),
                        "entity_type": str(chunk.get("entity_type", "")),
                        "tags": chunk.get("tags", []),
                    }, ensure_ascii=False),
                    now,
                    now,
                ),
            )
            inserted += 1
        except Exception as e:
            print(f"  skip chunk {i} ({str(chunk.get('id', ''))[:40]}): {e}")

    db.commit()
    db.close()

    print(f"Inserted {inserted}/{len(chunks)} chunks")
    print("Done.")


if __name__ == "__main__":
    main()
